File: Steffensen.cpp
double g(double x){
     return sin(x);
}


int main()
{

     double x1 = 1,tol = 1E-6,x2,x3,x4=0;
     int i = 1;
     while(abs(x4-x1) > tol){
          x4 = x1;
          x1 = x1 - (x2-x1)*(x2-x1)/(x3+x1-2*x2);
          x2 = g(x1);
          x3 = g(x2);
          i++;
     }
     cout<<i<<endl;
     printf("%.10f",x4);
     system("pause");
     return 0;
}

File: Steffensen.m
function xc = Steffensen(g,x0,tol)
x(1) = x0;
x(2) = g(x(1));
x(3) = g(x(2));
x(4) = x(3)-2*x(2)+x(1);
i = 1;
while abs(x(4)-x(1)) > tol
  x(4) = x(1);
  x(1) = x(3)-2*x(2)+x(1);
  x(2) = g(x(1));
  x(3) = g(x(2));
  i = i+1;
end
fprintf("经过%d次迭代得出",i)
xc = x(1);